File: src/DataScience/MATLAB/factory_method.m
function factory_method
%FACTORY_METHOD Keep useDatabase stable while varying the creation hook.
useDatabase(@createPostgres);
useDatabase(@createMySql);
end

function database = createPostgres
database = struct( ...
    'connect', @() fprintf('PostgreSQL connect\n'), ...
    'query', @() fprintf('PostgreSQL query\n'));
end

function database = createMySql
database = struct( ...
    'connect', @() fprintf('MySQL connect\n'), ...
    'query', @() fprintf('MySQL query\n'));
end

function useDatabase(createDatabase)
database = createDatabase();
database.connect();
database.query();
end

File: src/Other/MicroPython/factory_method.py
def create_postgres():
    return {
        "connect": lambda: print("PostgreSQL connect"),
        "query": lambda: print("PostgreSQL query"),
    }


def create_mysql():
    return {
        "connect": lambda: print("MySQL connect"),
        "query": lambda: print("MySQL query"),
    }


def use_database(create_database):
    database = create_database()
    database["connect"]()
    database["query"]()


use_database(create_postgres)
use_database(create_mysql)
